File: examples/kernel_ridge_ex.cpp
#include <iostream>
#include <random>
#include <cmath>

#include <Eigen/Dense>
#include "kernel_ridge.h"

using namespace std;

const double PI = 3.14159265358979323846;

double ftrue(double x) {
    return sin(2.0 * PI * x) + 0.5 * cos(4.0 * PI * x);
}

int main() {
    int n = 150;
    MatrixXd X(n, 1);
    VectorXd y(n);

    mt19937 gen(1);
    uniform_real_distribution<double> unif(-1.0, 1.0);
    normal_distribution<double> noise(0.0, 0.1);

    for (int i = 0; i < n; ++i) {
        double x = unif(gen);
        X(i, 0) = x;
        y(i) = ftrue(x) + noise(gen);
    }

    double lambda = 1e-3;
    double sigma = 0.2;

    KernelRidge model(lambda, sigma);
    model.fit(X, y);

    VectorXd fitted = model.predict(X);
    double train_mse = (fitted - y).squaredNorm() / y.size();

    cout << "Kernel Ridge Regression example" << endl;
    cout << "n = " << n << endl;
    cout << "lambda = " << lambda << endl;
    cout << "sigma = " << sigma << endl;
    cout << "y_mean = " << model.y_mean() << endl;
    cout << "Train MSE = " << train_mse << endl;

    int m = 10;
    MatrixXd X_grid(m, 1);

    for (int i = 0; i < m; ++i) {
        X_grid(i, 0) = -1.0 + 2.0 * i / (m - 1.0);
    }

    VectorXd pred = model.predict(X_grid);

    cout << "\nPredictions on grid:" << endl;
    for (int i = 0; i < m; ++i) {
        cout << "x = " << X_grid(i, 0)
             << ", yhat = " << pred(i) << endl;
    }

    return 0;
}

File: bindings/r/EigenRBinding/examples/kernel_ridge_ex.R
set.seed(1)

ftrue <- function(x) sin(2 * pi * x) + 0.5 * cos(4 * pi * x)

n <- 150
X <- matrix(0, n, 1)
y <- numeric(n)
for (i in 1:n) {
  x <- runif(1, -1, 1)
  X[i, 1] <- x
  y[i] <- ftrue(x) + rnorm(1, 0, 0.1)
}

lambda_ <- 1e-3
sigma <- 0.2

# construct and fit
kr <- new(KernelRidge, lambda_, sigma)
kr$fit(X, y)

fitted <- kr$predict(X)
train_mse <- mean((fitted - y)^2)

cat("Kernel Ridge Regression example\n")
cat("n =", n, "\n")
cat("lambda =", lambda_, "\n")
cat("sigma =", sigma, "\n")
cat("y_mean =", kr$y_mean(), "\n")
cat("Train MSE =", train_mse, "\n")

# prediction on a grid
m <- 10
X_grid <- matrix(seq(-1, 1, length.out = m), m, 1)
pred <- kr$predict(X_grid)

cat("\nPredictions on grid:\n")
for (i in 1:m) {
  cat(sprintf("x = %.4f, yhat = %.4f\n", X_grid[i, 1], pred[i]))
}

# plot: noisy data, true function, fitted curve
X_plot <- matrix(seq(-1, 1, length.out = 200), 200, 1)
y_true_plot <- ftrue(X_plot[, 1])
y_pred_plot <- kr$predict(X_plot)

plot(X[, 1], y, pch = 16, col = "grey60",
     xlab = "x", ylab = "y", main = "Kernel Ridge Regression")
lines(X_plot[, 1], y_true_plot, col = "black", lwd = 2, lty = 2)
lines(X_plot[, 1], y_pred_plot, col = "red", lwd = 2)
legend("topright",
       legend = c("noisy data", "true function", "KRR fit"),
       col = c("grey60", "black", "red"),
       pch = c(16, NA, NA),
       lty = c(NA, 2, 1),
       lwd = c(NA, 2, 2))
